Sources/GRPCHTTP2Core/GRPCMessageFramer.swift

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import NIOCore
+
+/// A ``GRPCMessageFramer`` helps with the framing of gRPC data frames:
+/// - It prepends data with the required metadata (compression flag and message length).
+/// - It compresses messages using the specified compression algorithm (if configured).
+/// - It coalesces multiple messages (appended into the `Framer` by calling ``append(_:compress:)``)
+/// into a single `ByteBuffer`.
+struct GRPCMessageFramer {
+  /// Length of the gRPC message header (1 compression byte, 4 bytes for the length).
+  static let metadataLength = 5
+
+  /// Maximum size the `writeBuffer` can be when concatenating multiple frames.
+  /// This limit will not be considered if only a single message/frame is written into the buffer, meaning
+  /// frames with messages over 64KB can still be written.
+  /// - Note: This is expressed as the power of 2 closer to 6KB (i.e., 6KiB) because `ByteBuffer`
+  /// reserves capacity in powers of 2. This way, we can take advantage of the whole buffer.
+  static let maximumWriteBufferLength = 16_384
+
+  private var pendingMessages: OneOrManyQueue<[UInt8]>
+
+  private var writeBuffer: ByteBuffer
+
+  private var compressor: Zlib.Compressor?
+
+  /// Create a new ``GRPCMessageFramer``.
+  /// - Parameter compressor: An optional compressor to use when compressing messages.
+  /// - Important: The `compressor` must have been `initialized()`.
```

Comment:
Hmm, the caller is also responsible for ending it (I assume) but the semantics here aren't clear because the framer holds on to it. I wonder if we should just pass it in to the `next` to make it clearer that the framer doesn't own the compressor. WDYT? (I appreciate that this requires bringing back the `PendingMessage`...)

Reply:
Ah well, this doc is outdated, forgot to update it - we don't need to `initialize` the compressor anymore, just end it. But yes, I agree the semantics aren't clear because we hold on to it... I think passing it in `next()` is not a bad idea. I don't think we need the `PendingMessage` wrapper though: we can make it an optional argument on `next()`, and compress if present, not compress if not.